dask/array/gufunc.py

```diff
@@ -0,0 +1,434 @@
+from __future__ import absolute_import, division, print_function
+
+import numpy as np
+from itertools import count
+import re
+
+try:
+    from cytoolz import concat, merge, unique
+except ImportError:
+    from toolz import concat, merge, unique
+
+from .core import Array, asarray, atop, getitem
+from .. import sharedict
+from ..core import flatten
+
+
+# Modified version of `numpy.lib.function_base._parse_gufunc_signature`
+# Modifications:
+#   - Allow for zero input arguments
+# See http://docs.scipy.org/doc/numpy/reference/c-api.generalized-ufuncs.html
+_DIMENSION_NAME = r'\w+'
+_CORE_DIMENSION_LIST = '(?:{0:}(?:,{0:})*,?)?'.format(_DIMENSION_NAME)
+_ARGUMENT = r'\({}\)'.format(_CORE_DIMENSION_LIST)
+_INPUT_ARGUMENTS = '(?:{0:}(?:,{0:})*,?)?'.format(_ARGUMENT)
+_OUTPUT_ARGUMENTS = '{0:}(?:,{0:})*'.format(_ARGUMENT)  # Use `'{0:}(?:,{0:})*,?'` if gufunc-
+# signature should be allowed for length 1 tuple returns
+_SIGNATURE = '^{0:}->{1:}$'.format(_INPUT_ARGUMENTS, _OUTPUT_ARGUMENTS)
+
+
+def _parse_gufunc_signature(signature):
+    """
+    Parse string signatures for a generalized universal function.
+
+    Arguments
+    ---------
+    signature : string
+        Generalized universal function signature, e.g., ``(m,n),(n,p)->(m,p)``
+        for ``np.matmul``.
+
+    Returns
+    -------
+    Tuple of input and output core dimensions parsed from the signature, each
+    of the form List[Tuple[str, ...]], except for one output. For one  output
+    core dimension is not a list, but of the form Tuple[str, ...]
+    """
+    signature = signature.replace(' ', '')
+    if not re.match(_SIGNATURE, signature):
+        raise ValueError('Not a valid gufunc signature: {}'.format(signature))
+    in_txt, out_txt = signature.split('->')
+    ins = [tuple(re.findall(_DIMENSION_NAME, arg))
+           for arg in re.findall(_ARGUMENT, in_txt)]
+    outs = [tuple(re.findall(_DIMENSION_NAME, arg))
+            for arg in re.findall(_ARGUMENT, out_txt)]
+    outs = outs[0] if ((len(outs) == 1) and (out_txt[-1] != ',')) else outs
```

Comment:
Why does this have a different return type for a single output? I think that makes your logic below unnecessarily complicated.